### ISSUE TITLE
Invalid HTTP response can crash Guzzle-Event-Loop

### ISSUE THREAD

**Issue by not-implemented** [guzzle/guzzle#2266]
**Guzzle version(s) affected**: 6.5.1

**Description**  
If a (broken) server responds with a "HTTP/1.1 404: Not Found" header (note the wrong colon!) **after being redirected**, the promise is not normally rejected, but the `$promise->wait();` call throws an Exception ("Status code must be an integer value.") - and exits the event-loop completely (very bad when having concurrent requests).

**How to reproduce**  
Start two "HTTP test servers" ...

... one on port 18000 that redirects to port 18001:
```
cat <<EOF | nc -N -l -p 18000
HTTP/1.1 301 Moved Permanently
Connection: close
Location: http://localhost:18001/

EOF
```

... and one on port 18001 that sends the broken header with the colon after "404":
```
cat <<EOF | nc -N -l -p 18001
HTTP/1.1 404: Not Found
Connection: close

EOF
```

... then execute this script:
```php
$url = 'http://localhost:18000/';

try {
    $client = new \GuzzleHttp\Client();
    $client->requestAsync('GET', $url)->then(function ($response) {
        echo 'Response Status-Code: ' . $response->getStatusCode() . PHP_EOL;
    }, function ($e) {
        echo 'Normal Error: ' . $e->getMessage() . PHP_EOL;
    })->wait();
} catch (\Exception $e) {
    echo 'Unexpected Error (should never happen): ' . $e->getMessage() . PHP_EOL;
}
```

It will run into the "Unexpected Error"-catch.

**Possible Solution**  
The problem is, that `$easy->createResponse();` in https://github.com/guzzle/guzzle/blob/master/src/Handler/CurlFactory.php#L569 can throw exceptions, but is not inside a try-catch.

Maybe it is the best, to just expand the try-catch for "$onHeaders"-callback so that the `$easy->createResponse();` is covered, too?

**Comment by edmondscommerce** [comment]
Also seeing this issue

**Comment by bizmate** [comment]
@not-implemented have you found a workaround for this? 
Even if I `try/catch` the exception the code still exits

```
try{
	$results = GuzzlePromise\settle($requestPromises)->wait();
}
catch(Exception $e){
	$msg = __METHOD__ . ' : exception for sites: ' . var_export( $sitesToCheck, true ) . ' exception: ' .
	get_class($e) . ' code: ' . $e->getCode() . ' message: ' . $e->getMessage() . ' trace: ' .
	$e->getTraceAsString();
			
	$this->logger->err( $msg );
}
```

Output
```
  [InvalidArgumentException]                                 
  Status code must be an integer value between 1xx and 5xx.  
                                                             

Exception trace:
 () at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/psr7/src/Response.php:151
 GuzzleHttp\Psr7\Response->assertStatusCodeRange() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/psr7/src/Response.php:98
 GuzzleHttp\Psr7\Response->__construct() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/guzzle/src/Handler/EasyHandle.php:81
 GuzzleHttp\Handler\EasyHandle->createResponse() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:560
 GuzzleHttp\Handler\CurlFactory->GuzzleHttp\Handler\{closure}() at n/a:n/a
 curl_multi_exec() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:114
 GuzzleHttp\Handler\CurlMultiHandler->tick() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:131
 GuzzleHttp\Handler\CurlMultiHandler->execute() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:246
 GuzzleHttp\Promise\Promise->invokeWaitFn() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:223
 GuzzleHttp\Promise\Promise->waitIfPending() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:267
 GuzzleHttp\Promise\Promise->invokeWaitList() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:225
 GuzzleHttp\Promise\Promise->waitIfPending() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:62
 GuzzleHttp\Promise\Promise->wait() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/EachPromise.php:101
 GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure}() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:246
 GuzzleHttp\Promise\Promise->invokeWaitFn() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:223
 GuzzleHttp\Promise\Promise->waitIfPending() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:267
 GuzzleHttp\Promise\Promise->invokeWaitList() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:225
 GuzzleHttp\Promise\Promise->waitIfPending() at /var/www/sitecrawler/symfony3/vendor/guzzlehttp/promises/src/Promise.php:62
```

and `$this->logger->err( $msg );` is never called

**Comment by bizmate** [comment]
Another quick update on this, i see the problem when a standard 303 response is received. I guess guzzle does not handle it well?

Curl call returns a standard 303 as shown below. I wonder why Guzzle just panics on this.
```
curl -v bandwidth-infr******re.com
* Rebuilt URL to: bandwidth-infr******re.com/
*   Trying 64.***.30...
* TCP_NODELAY set
* Connected to bandwidth-infr******re.com (64.***.30) port 80 (#0)
> GET / HTTP/1.1
> Host: bandwidth-infr******re.com
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 303 See Other
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Status: 303 See Other
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< Location: http://www.linkedin.com/in/e******a
< Cache-Control: no-cache
< X-Request-Id: 71b5c939-55fe-4416-8987-ce20f77a74ff
< X-Runtime: 0.004081
< X-Powered-By: Phusion Passenger 4.0.53
< Date: Thu, 05 Dec 2019 12:03:39 GMT
< Server: nginx/1.6.2 + Phusion Passenger 4.0.53
< P3P: CP="IDC DSP COR ADM DEVi TAIi PSA PSD IVAi IVDi CONi HIS OUR IND CNT"
< 
* Connection #0 to host bandwidth-infrastructure.com left intact
<html><body>You are being <a href="http://www.linkedin.com/in/e*******a">redirected</a>.</body></html>
```

**Comment by not-implemented** [comment]
@bizmate No I do not have a workaround for this :-(

**Comment by not-implemented** [comment]
I updated the "reproduce steps" of the issue and the "possible solution".

Maybe one of the core developers can have a look at it? :-) Maybe @gmponos ?

**Comment by not-implemented** [comment]
#2591 may fix this

**Comment by Nyholm** [comment]
This is fixed in 7.2.0
